### ISSUE TITLE
NIST 800-53r4 - AC-5 - Separation of Duties

### ISSUE THREAD

**Issue by ruckc** [ozoneplatform/owf-framework#75]
Currently Ozone 7 requires all users (including Admins) to have the User role.  This a little hard as RMF requirements for AC-5 require a separation in roles and responsibilities.  Ideally this check should be checking if UserIsUser **OR** if UserIsAdmin.  It is also quite frustrating that this logic is hardcoded instead of being placed in the spring security configuration.

``` 
           if (!accountService.getLoggedInUserIsUser()) {
	                    log.debug(accountService.getLoggedInUsername() + " does not have ROLE_USER role, erroring out with 401");
	                    response.sendError(401)
	                    return false
	       }
```

**Comment by rpokorny** [comment]
The workflows in OWF are such that an administrative user must have a superset of the capabilities that a normal user has.  There are two ways that we could have codified this in the roles system:

1. ROLE_USER has capabilities x, y, and z, and ROLE_ADMIN has x, y, z, and q (ROLE_ADMIN implicitly has every capability that ROLE_USER has and more).
2. ROLE_USER has capabilities x, y, and z and ROLE_ADMIN has capability q.  In order to actually use the system in a meaningful way, a user with ROLE_ADMIN must also have ROLE_USER since x, y and z are necessary for basic system usage.

Of these two options, we chose the latter.  From a security perspective, they are equivalent - normal users have capabilities x, y, and z and administrative users have capabilities x, y, z, and q.  I do not believe that option 2 from the list above is any less appropriate for meeting AC-5 than option 1 would be.  

I don't completely follow what you are saying about spring security.  A user who has ROLE_ADMIN and not ROLE_USER cannot effectively use the app.  The code snippet that you posted was put in place to make that clear up front.  Having a configuration to disable that check would only let those users into the app in a state where things would be confusingly nonfunctional.

**Comment by ruckc** [comment]
So, ROLE_USER infers that the user gets all widgets granted to that role.  Those widgets are typically for regular system use, not administrative.  The ADMIN, when logged in as admin, shouldn't be using the regular system, they should be managing the system then logging out.  By forcing ADMINs to be a member of USERs, allows the admins to lazily do regular system usage, when they are overprivileged.

Essentially, in my view, administrators typically have separate accounts these days, one for administrative work and one for user-level non-administrative work.  In Microsoft Windows, you don't typically want your Administrators running Outlook and checking their email, as that doesn't require administrative privileges, and by them running outlook as an administrator, exposes their system and potentially the network to a much higher risk.

Essentially, the ROLE_USER should only have the ability to see/open apps/stacks/ and app components, and ROLE_ADMIN should only have the ability to manage OWF's apps, stacks, and app components; not necessarily use them.

**Comment by rpokorny** [comment]
I think the access controls that you desire can be handled by Groups.  A user can be in any number of Groups, which can be assigned by an administrator or declared via the security plugin.  Additionally there are two "special" groups - OWF Users and OWF Admins, which are specially coded to map to ROLE_USER and ROLE_ADMIN respectively.    Apps and App Components can be associated with a Group, making them accessible to all Users within that Group.  

An administrative user is going be a member of at least OWF Users and OWF Admins, but they aren't necessarily going to be a member of anything else.  Therefore you can prevent administrators from accessing certain components by putting those components in some other group which only your "normal" users are a part of.  There are two caveats here:

1. Administrators are of course the ones that are in charge of app component <-> group assignments, so there would be nothing preventing them from undoing this arrangement
2. App Components are distinct web apps.  Anyone can access an app component by typing its URL into their browser address bar, regardless of what access they have to OWF (if any).  As a result, the group management system in OWF is really more of a UI convenience than an actual security safeguard - all App Components must implement their own auth&auth separately from OWF's user management.

**Comment by ruckc** [comment]
If you actually rearranged the logic to it would allow for admins to exist without ROLE_USER.

           if (!accountService.getLoggedInUserIsUser() && !accountService.getLoggedInUserIsAdmin()) {
                        log.debug(accountService.getLoggedInUsername() + " does not have ROLE_USER role, erroring out with 401");
                        response.sendError(401)
                        return false
           }

**Comment by rpokorny** [comment]
It would allow them to get past that spot, yes.  But it wouldn't allow them to function meaningfully elsewhere in the app.  This code snipped exists _specifically_ to prevent people with ROLE_ADMIN but not ROLE_USER from logging into the system.  It is a fail-fast safeguard so that the user gets a clear message of what's wrong instead of logging in successfully but then having all sorts of necessary operations confusingly fail.

**Comment by ruckc** [comment]
Given the rest of the OWF environment, to my understanding it would allow a ROLE_ADMIN into OWF, but they would only have access to the Administration App/Stacks/Widgets.  To me this seems like the right answer.

**Comment by rpokorny** [comment]
Please see my comment two comments ago.  If you are only worried about limiting ROLE_ADMIN access to the administrative Apps and App Components, then the existing Groups mechanism can easily achieve that.

**Comment by ruckc** [comment]
I've never seen other groups actually work, the "magic" to tie OWF_Users and OWF_Admins to ROLE_USER/ADMIN seems to be the only automatic membership groups that have ever worked, and not having automatic groups is a non-starter.

**Comment by rpokorny** [comment]
If you believe automatic groups are broken please open a ticket for that with Steps to Reproduce.  To my knowledge they are working.

**Comment by rpokorny** [comment]
Actually, I take that back.  I had fixed some issues with automatic groups fairly recently in commits a6b5ceb0574d3 and 28b517a554be.  If you are seeing problems with automatic groups in 7.17.0, please try building master and seeing if that the issue is fixed.  Those commits will be included in an upcoming 7.17.1 release.
